lldb/tools/lldb-dap/DAP.cpp

```diff
@@ -671,8 +695,33 @@ void DAP::SetTarget(const lldb::SBTarget target) {
   }
 }
 
-bool DAP::HandleObject(const protocol::Message &M) {
-  if (const auto *req = std::get_if<protocol::Request>(&M)) {
+bool DAP::HandleObject(const Message &M) {
+  if (const auto *req = std::get_if<Request>(&M)) {
+    {
+      std::lock_guard<std::mutex> lock(m_active_request_mutex);
+      m_active_request = req;
+    }
+
+    auto cleanup = llvm::make_scope_exit([&]() {
+      std::scoped_lock<std::mutex> active_request_lock(m_active_request_mutex);
+      m_active_request = nullptr;
+    });
+
+    {
+      // If there is a pending cancelled request, preempt the request and mark
+      // it cancelled.
+      std::lock_guard<std::mutex> lock(m_cancelled_requests_mutex);
+      if (m_cancelled_requests.find(req->seq) != m_cancelled_requests.end()) {
+        Response cancelled = CancelledResponse(req->seq, req->command);
+        Send(cancelled);
+        return true;
+      }
+    }
+
+    // Clear interrupt marker prior to handling the next request.
+    if (debugger.InterruptRequested())
+      debugger.CancelInterruptRequest();
```

Comment:
You still have the same race here. The request could be cancelled between this line and when the active request field is set on line 702. (My complaint wasn't so much about the exact std::atomic type as it was about treating this flag as an atomic -- a mutex surrounding just a variable assignment is morally equivalent to an atomic). For this to work, I think the flag needs to be checked in the same critical section that handles  sets `m_active_request`. I believe it will be fine if you move this code to line 702, just before setting the active request.

Reply:
Moved this up to 702